Support doctrine/mongodb-odm ^2.x & doctrine/mongodb-odm-bundle ^4.x

romainnorberg opened the pull request:
- support both version of `doctrine/mongodb-odm` (`^1.0 || ^2.0`)
    - https://github.com/doctrine/mongodb-odm/blob/master/UPGRADE-2.0.md
    - **BC**: https://github.com/doctrine/mongodb-odm/pull/2069
- support both version of  `doctrine/mongodb-odm-bundle` (`^3.0 || ^4.0`) 
    - https://github.com/doctrine/DoctrineMongoDBBundle/blob/master/UPGRADE-4.0.md
- run php-cs fixer _(lint CI build)_
- updated `FilterTypeGuesserTest`

<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 3.x is for everything backwards compatible, like patches, features and deprecation notices
    - master is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/blob/3.x/CONTRIBUTING.md#the-base-branch
-->
I am targeting this branch, because this is not BC.

core23 commented:
AFAIK you could do this on the stable branch. 

WDYT @sonata-project/contributors

core23 commented on src/Guesser/FilterTypeGuesser.php:
```diff
@@ -49,10 +49,10 @@ public function guessType($class, $property, ModelManagerInterface $modelManager
-                    //case ClassMetadataInfo::MANY_TO_MANY:
+                case ClassMetadata::ONE:
+                case ClassMetadata::MANY:
+                    //case ClassMetadata::MANY_TO_ONE:
```
This comments are superfluous, you can remove them

OskarStark commented:
Yes let’s do it on the stable branch

romainnorberg commented:
hello @core23 @OskarStark Thanks for reviews 👍 

I don't have a lot of experience in BC management _(I thought I had to target the master branch with these changes)_ and I am a novice in the contribution to open-source packages 😕 

I'll check why CI fail

aturki commented:
Hello,
Both doctrine/mongodb-odm and doctrine/mongodb-odm-bundle require PHP version 7.2 or higher. This is a BC break with the current version 3.x. Shouldn't we target the master branch instead ?

phansys commented:
IMO, from a strict semver point of view, if the changes are respecting BC we could accept them regardless their dependencies; since PHP is in fact a dependency itself.
But we have some [extra rules](https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/blob/3.x/CONTRIBUTING.md#dependency-changes) around it which disallow the drop of a security maintained PHP version on a stable branch.

core23 commented on src/Builder/FormContractor.php:
```diff
@@ -13,7 +13,7 @@
 namespace Sonata\DoctrineMongoDBAdminBundle\Builder;
 
-use Doctrine\ODM\MongoDB\Mapping\ClassMetadataInfo;
+use Doctrine\ODM\MongoDB\Mapping\ClassMetadata;
```
This class exists in odm 1.x and 2.x. You could support both version inside the composer.json file

romainnorberg commented:
hello @core23 thanks for review.

- I updated composer.json file to support both version of mongodb-odm
- run php-cs fixer _(lint CI build)_
- updated `FilterTypeGuesserTest` -> i don't know if it's the right solution 👀

greg0ire commented on composer.json:
```diff
@@ -26,8 +26,8 @@
-        "doctrine/mongodb-odm": "^1.0",
-        "doctrine/mongodb-odm-bundle": "^3.0",
+        "doctrine/mongodb-odm": "^1.0 || ^2.0",
+        "doctrine/mongodb-odm-bundle": "^4.0",
```
Why are we not using `^3.0 || ^4.0` here? No way the 7.1 build works otherwise…

romainnorberg replied:
hello @greg0ire you're right, I updated the composer.json. thank u for the review

greg0ire commented on tests/Guesser/FilterTypeGuesserTest.php:
```diff
@@ -35,6 +35,9 @@ public function testThrowsOnMissingField()
         $property = 'whatever';
+
+        $this->metadata->hasAssociation($property)->willReturn(false);
+
```
Why did you add this line exactly?

romainnorberg replied:
method `metadata->hasAssociation` must return a boolean _(typed since https://github.com/doctrine/mongodb-odm/pull/1846 I think)_ Without this line, test fail.


Test log https://travis-ci.org/sonata-project/SonataDoctrineMongoDBAdminBundle/jobs/593439895#L498

<details>
1) Sonata\DoctrineMongoDBAdminBundle\Tests\Guesser\FilterTypeGuesserTest::testThrowsOnMissingField
Failed asserting that exception of type "TypeError" matches expected exception "Sonata\DoctrineMongoDBAdminBundle\Model\MissingPropertyMetadataException". Message was: "Return value of Double\Doctrine\ODM\MongoDB\Mapping\ClassMetadata\P8::hasAssociation() must be of the type bool, null returned" at
/home/travis/build/sonata-project/SonataDoctrineMongoDBAdminBundle/src/Guesser/FilterTypeGuesser.php:48
/home/travis/build/sonata-project/SonataDoctrineMongoDBAdminBundle/tests/Guesser/FilterTypeGuesserTest.php:43
</details>

cf:
- https://github.com/romainnorberg/SonataDoctrineMongoDBAdminBundle/blob/master/src/Guesser/FilterTypeGuesser.php#L48

greg0ire replied:
Oh ok, there was probably no type declaration before in `ClassMetadata`, so it was fine with returning `null`. I think your fix is ok 👍

romainnorberg commented:
thanks @all for your time & reviews

OskarStark commented:
We thank you for this contribution, keep on going 👍🏻👍🏻🥳🥳
